serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,282 @@
+		return ErrNamespaceUnset
+	}
+	if podName == "" {
+		return ErrPodNameUnset
+	}
+	if len(patches) == 0 {
+		// No work to perform.
+		return nil
+	}
+
+	var jsonPatches []interface{}
+	for _, patch := range patches {
+		if patch.Operation == Unset {
+			return errors.New("patch operation must be set")
+		}
+		jsonPatches = append(jsonPatches, map[string]interface{}{
+			"op":    patch.Operation.String(),
+			"path":  patch.Path,
+			"value": patch.Value,
+		})
+	}
+	body, err := json.Marshal(jsonPatches)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest(method, c.config.Host+endpoint, bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json-patch+json")
+	return c.do(req, nil)
+}
+
+// do executes the given request, retrying if necessary.
+func (c *Client) do(req *http.Request, ptrToReturnObj interface{}) error {
+	// Finish setting up a valid request.
+	req.Header.Set("Authorization", "Bearer "+c.config.BearerToken)
+	req.Header.Set("Accept", "application/json")
+	client := cleanhttp.DefaultClient()
+	client.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: c.config.CACertPool,
+		},
+	}
+
+	// Execute and retry the request. This exponential backoff comes
+	// with jitter already rolled in.
+	var lastErr error
+	b := backoff.NewExponentialBackOff()
+	for i := 0; i < maxRetries; i++ {
+		if i != 0 {
+			select {
+			case <-c.stopCh:
+				return nil
+			case <-time.NewTimer(b.NextBackOff()).C:
+				// Continue to the request.
+			}
+		}
+		shouldRetry, err := c.attemptRequest(client, req, ptrToReturnObj)
+		if !shouldRetry {
+			// The error may be nil or populated depending on whether the
+			// request was successful.
+			return err
+		}
+		lastErr = err
+	}
+	return lastErr
+}
+
+// attemptRequest tries one single request. It's in its own function so each
+// response body can be closed before returning, which would read awkwardly if
+// executed in a loop.
+func (c *Client) attemptRequest(client *http.Client, req *http.Request, ptrToReturnObj interface{}) (shouldRetry bool, err error) {
+	// Preserve the original request body so it can be viewed for debugging if needed.
+	// Reading it empties it, so we need to re-add it afterwards.
+	var reqBody []byte
+	if req.Body != nil {
+		reqBody, _ = ioutil.ReadAll(req.Body)
+		reqBodyReader := bytes.NewReader(reqBody)
+		req.Body = ioutil.NopCloser(reqBodyReader)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			if c.logger.IsWarn() {
+				// Failing to close response bodies can present as a memory leak so it's
+				// important to surface it.
+				c.logger.Warn(fmt.Sprintf("unable to close response body: %s", err))
+			}
+		}
+	}()
+
+	// Check for success.
+	switch resp.StatusCode {
+	case 200, 201, 202:
+		// Pass.
+	case 401, 403:
+		// Perhaps the token from our bearer token file has been refreshed.
```

Comment:
Really nice feature here!

Reply:
Thanks! :-) I thought about checking it periodically but I couldn't find a document about how often it changes, so decided to do it lazily instead.